course-material/intro-to-sql/aggregate-functions-in-sql.md

```diff
@@ -1,14 +1,287 @@
 ---
-jupyter:
-  jupytext:
-    text_representation:
-      extension: .md
-      format_name: markdown
-      format_version: '1.3'
-      jupytext_version: 1.14.5
+jupytext:
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.5
+kernelspec:
+  display_name: jupyblog
+  language: python
+  name: python3
 ---
 
+
 # Aggregate functions in SQL
 
+In the last section, we learned the basics of SQL. We now build upon those basics to learn a common use case of SQL: <b>aggregation functions</b>. 
+
+Aggregation functions are useful for summarizing your data and for finding meaningful insights. The most common of these functions are `COUNT()`, `AVG()`, `SUM()`, `MIN()`, and `MAX()`. We will go over each of these functions in detail with our bank marketing dataset.
+
+We will also introduce the important clause `GROUP BY`. This clause can only be used if, and only if, aggregation functions are used. However, note that aggregation functions can be used without `GROUP BY`. `GROUP BY` groups unique values in columns together and runs an aggregation function on each unique group. Examples of using `GROUP BY` will be provided after introducing aggregate functions.
+
+Let's first run the installations and setup before running any queries, just like the last lesson `making-your-first-query`.
+
+<!-- region -->
+
+### Install - execute this once. 
+<b>Note:</b> If you are following these lessons locally and <b>not</b> on Google Colab, then there is no need to reinstall these packages.
+
+This code installs JupySQL, DuckDB, and Pandas in your environment. We will be using these moving forward.
+
+```{code-cell} ipython3
+%pip install jupysql --upgrade duckdb-engine pandas --quiet
+```
+
+### Load the data
+<b>Note:</b> If you are following these lessons locally and <b>not</b> on Google Colab, then there is no need to load the data again. 
+
+We extract the bank marketing data by retrieving it from it's URL download link. The link may be a zip file (which it is in this case), so we extract the zip file, read the file containing the data within the zip file, and clean the data. Finally, we save this cleaned data to it's own seperate file called `bank_cleaned.csv`.  
+```python
+from urllib.request import urlretrieve
+from zipfile import ZipFile
+import pandas as pd
+import os
+
+def extract_to_csv(url, data_name):
+  #Retrieve the zip file from the url link
+  file = os.path.basename(url)
+  urlretrieve(url, file)
+
+  #Extract the zip file's contents
+  with ZipFile(file, 'r') as zf:
+    zf.extractall()
+
+  #The file containing our data
+  csv_file_name = f'{data_name}.csv'
+
+  # Data clean up
+  df = pd.read_csv(csv_file_name, sep = ";")
+
+  # Save the cleaned up CSV file
+  df.to_csv(df.to_csv(f'{data_name}.csv', index=False))
+
+#Running the above function
+extract_to_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/00222/bank.zip', bank)
```

Comment:
This will return an error message as the variable bank is not defined. 

Option 1: define

`bank = 'bank'`

option 2: pass the string 'bank' to the function